src/kvstore/raftex/RaftPart.cpp

```diff
@@ -186,17 +135,172 @@ class AppendLogsIterator final : public LogIterator {
 
  private:
   size_t idx_{0};
-  bool leadByAtomicOp_{false};
-  bool hasNonAtomicOpLogs_{false};
-  bool valid_{true};
-  LogType lastLogType_{LogType::NORMAL};
   LogType currLogType_{LogType::NORMAL};
-  std::optional<std::string> opResult_;
   LogID firstLogId_;
   TermID termId_;
   LogID logId_;
   RaftPart::LogCache logs_;
-  OpProcessor opCB_;
+};
+
+class AppendLogsIteratorFactory {
+ public:
+  AppendLogsIteratorFactory() = default;
+  void make(RaftPart::LogCache& cacheLogs, RaftPart::LogCache& sendLogs) {
+    for (auto& log : cacheLogs) {
+      auto code = mergeAble(log);
+      if (code == MergeAbleCode::MERGE_BOTH) {
+        sendLogs.emplace_back();
+        std::swap(cacheLogs.front(), sendLogs.back());
+        cacheLogs.pop_front();
+        continue;
+      } else if (code == MergeAbleCode::MERGE_PREV) {
+        sendLogs.emplace_back();
+        std::swap(cacheLogs.front(), sendLogs.back());
+        cacheLogs.pop_front();
+        break;
+      } else {  // NONE, or MERGE_NEXT
+        break;
+      }
+    }
+  }
+
+  /**
+   * @brief check if a incoming log can be merged with previous logs
+   *
+   * @param logWrapper
+   */
+  MergeAbleCode mergeAble(RaftPart::LogCacheItem& logWrapper) {
+    // log type:
+    switch (std::get<1>(logWrapper)) {
+      case LogType::NORMAL: {
+        std::vector<folly::StringPiece> updateSet;
+        auto& log = std::get<2>(logWrapper);
+        if (log.empty()) {
+          return MergeAbleCode::MERGE_BOTH;
+        }
+        decode(log, updateSet, ranges_);
+        for (auto& key : updateSet) {
+          memLock_.insert(key.str());
+        }
+        return MergeAbleCode::MERGE_BOTH;
+      }
+      case LogType::COMMAND: {
+        return MergeAbleCode::MERGE_PREV;
+      }
+      case LogType::ATOMIC_OP: {
+        auto& atomOp = std::get<3>(logWrapper);
+        auto [code, result, read, write] = atomOp();
+        if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+          DLOG(INFO) << "===> OOPs, atomOp failed!!!, code = "
+                     << apache::thrift::util::enumNameSafe(code);
+          auto& promiseRef = std::get<4>(logWrapper);
+          if (!promiseRef.isFulfilled()) {
+            promiseRef.setValue(code);
+          }
+          return MergeAbleCode::MERGE_BOTH;
+        }
+        std::get<2>(logWrapper) = std::move(result);
+        /**
+         * @brief We accept same read/write key in a one log,
+         *        but reject if same in different logs.
+         */
+        for (auto& key : read) {
+          auto cit = memLock_.find(key);
+          // read after write is not acceptable.
+          if (cit != memLock_.end()) {
+            return MergeAbleCode::MERGE_NEXT;
```

Comment:
Where do we clear `memLock_`?

Reply:
Change AppendIteratorFactory::make() to static. MemLock will release after exit that function.